### DISCUSSION TITLE
Sorting and Filtering

### DISCUSSION THREAD

**Discussion by Merijeek** [advplyr/audiobookshelf#16] [Ideas]
Looking at the V1.0 right now, and adding and messing with books, and I'm seeing a couple things. 

1. If I do a sort by Author (last, first) it works. But I can't figure out how it is ordering them after that. So all the books by Smtih, John seem to be sorted...randomly (?) within the Smith, John bucket. 

2. Having a search up top is nice, but the ability to filter would be nice, too. Just a filter field where I can put in "smith" and it will locate anything with "smith" in any of the fields.

**Comment by advplyr** [comment]
> If I do a sort by Author (last, first) it works. But I can't figure out how it is ordering them after that. So all the books by Smtih, John seem to be sorted...randomly (?) within the Smith, John bucket.

It would be ordering them by when they were added. The difficulty with supporting multi-level sorting would be the user interface.

> Having a search up top is nice, but the ability to filter would be nice, too. Just a filter field where I can put in "smith" and it will locate anything with "smith" in any of the fields.

Yeah this is a good idea, something like "filter by keyword". It may also need to include a dropdown of which fields to search for the keyword in, maybe you put something in the description of each book and you want to filter for that.

It also needs a filter by author and the genre filter should only list the genres that are being used.

The entire filter & sort needs to be re-thought out. Right now it persists the filter so if you logout and log back in your filter should remain. Persisting author names and user-entered keywords will get complicated, so I'm considering only persisting the sort option.

**Comment by advplyr** [comment]
I re-thought out the filtering. Unfortunately this is going to impact existing genres and tags. The genres and tags already assigned will remain, but they will be in snake_case.

This change will make it possible to use any UTF-8 character in a stored filter. Filters are no longer human-readable when stored, they are stored as Base64 strings.

For that reason I was able to add Authors to the filter dropdown. v1.0.3 also includes a keyword filter that doesn't persist. I don't think that keyword filter will stay like that, but I couldn't come up with a clean way to interface it.
